nums/core/linalg.py

```diff
@@ -256,6 +256,147 @@ def inv(app: ArrayApplication, X: BlockArray):
+
+        # Collect data for R01
+        for inc in range(R01_num_blocks):
+            R01_oids.append(R.blocks[(R01_sb_row + inc, R01_sb_col)].oid)
+            R01_shapes.append(R.blocks[(R01_sb_row + inc, R01_sb_col)].shape)
+            R01_grid_entries.append((R01_sb_row + inc, R01_sb_col))
+
+        # Perform matrix multiplication: R01_1 = -R00 @ R01
+        R01_1_oids = []
+        for row_block in range(R01_num_blocks):
+            sub_oids = []
+
+            for col_block in range(R01_num_blocks):
+
+                # Get data for R00
+                R00_oid = R.blocks[(row_block, col_block)].oid
+                Z_oid = Zs.blocks[(row_block, col_block)].oid
+
+                # Calculate -R00 = 0 - R00
+                neg_R00_oid = app.cm.bop(
+                    "subtract",
+                    Z_oid,
+                    R00_oid,
+                    False,
+                    False,
+                    axes=1,
+                    syskwargs={
+                        "grid_entry": (row_block, col_block),
+                        "grid_shape": grid_shape,
+                    },
+                )
+                # Calculate -R00 @ R01
+                sub_oids.append(
+                    app.cm.bop(
+                        "tensordot",
+                        neg_R00_oid,
+                        R01_oids[col_block],
+                        False,
+                        False,
+                        axes=1,
+                        syskwargs={
+                            "grid_entry": R01_grid_entries[col_block],
+                            "grid_shape": grid_shape,
+                        },
+                    )
+                )
+
+            # Finished with one blocked mult
+            R01_1_oids.append(
+                app.cm.sum_reduce(
+                    *sub_oids,
+                    syskwargs={
+                        "grid_entry": R01_grid_entries[row_block],
+                        "grid_shape": grid_shape,
+                    }
+                )
+            )
+
+        # Perform matrix multiplication: R_01_2 = R_01_1 @ R_11_inv
+        R01_2_oids = []
+        for row_block in range(R01_num_blocks):
+            R01_2_oids.append(
+                app.cm.bop(
+                    "tensordot",
+                    R01_1_oids[row_block],
+                    R11_inv_oid,
+                    False,
+                    False,
+                    axes=1,
+                    syskwargs={
+                        "grid_entry": R01_grid_entries[row_block],
+                        "grid_shape": grid_shape,
+                    },
+                )
+            )
+
+        # Reset R_01
+        for i, entry in enumerate(R01_grid_entries):
+            R.blocks[entry].oid = R01_2_oids[i]
+
+        # Recompute R_tl.shape
+        r11_r, r11_c = R11_shape
+        old_r, old_c = R_tl_shape
+
+        R_tl_shape = (old_r + r11_r, old_c + r11_c)
+
+    # By the time we finish, R = R_inv
```

Comment:
Before returning, I think we'll need to reshape `R` to match the block shape of the input array `X`.